crates/interpreter/src/instructions/system.rs

```diff
@@ -179,6 +157,29 @@ pub fn returndataload<H: Host + ?Sized>(interpreter: &mut Interpreter, _host: &m
     *offset = B256::from(output).into();
 }
 
+// common logic for copying data from a source buffer to the EVM's memory
+fn copy_to_memory(
+    interpreter: &mut Interpreter,
+    memory_offset: U256,
+    data_offset: usize,
+    len: usize,
+    source: *const [u8],
+) {
+    gas_or_fail!(interpreter, gas::verylowcopy_cost(len as u64));
+    if len == 0 {
+        return;
+    }
+    let memory_offset = as_usize_or_fail!(interpreter, memory_offset);
+    resize_memory!(interpreter, memory_offset, len);
+
+    // Note: this can't panic because we resized memory to fit.
+    unsafe {
+        interpreter
+            .shared_memory
+            .set_data(memory_offset, data_offset, len, &*source);
+    }
+}
```

Comment:
I would remove this, and change behaviour of function to do gas, len check, memory resize.

As memory_offset is reused, we can return that from this function as `Option<usize>`

Reply:
The problem with changing the functions behaviour as suggested by you is that we encountering is a classic case of Rust's borrow checker preventing simultaneous mutable and immutable borrows of the same data. 
When we do :
```rust
let source = interpreter.contract.input.as_ref();
    let memory_offset = copy_to_memory(interpreter, memory_offset, data_offset, len);
    if let Some(memory_offset) = memory_offset {
        // Note: this can't panic because we resized memory to fit.
        interpreter.shared_memory.set_data(memory_offset, data_offset, len, source);
    };
```
We create an immutable borrow of interpreter.contract.input and then we try to mutably borrow interpreter in the copy_to_memory function call. Finally, we try to use the immutable borrow source again in the set_data call.
The Rust borrow checker doesn't allow this because it can't guarantee that the mutable borrow in copy_to_memory won't invalidate the source reference.

How do you suggest I tackle this! @rakita

Reply:
can this line: `let source = interpreter.contract.input.as_ref();` be moved inside `if let Some(memory_..` so you don't have a borrow?

Reply:
Now when I return `Option<usize>` I get `expected enum `std::option::Option<usize>`
found unit type `()`` from the gas! macro. Thats why the tests are failing. How do you suggest I tackle this! @rakita

Reply:
* There is a gas! version that takes Return. `gas!(10,None)` something similar should be added for `gas_or_fail!`
* Change `as_usize_or_fail` to `as_usize_or_fail_ret!(interpreter, memory_offset,None);`
* Change `resize_memory!` to `resize_memory!(interpreter, memory_offset, len, None);`